src/ircode.c

```diff
@@ -788,7 +788,7 @@ JL_DLLEXPORT jl_code_info_t *jl_uncompress_ir(jl_method_t *m, jl_code_instance_t
 
     jl_code_info_t *code = jl_new_code_info_uninit();
     uint8_t flags = read_uint8(s.s);
-    code->aggressive_constprop = !!(flags & (1 << 4));
+    code->constprop = (flags >> 4) & 3;
```

Comment:
I may miss some basic understanding, but why do we need `& 3` here (isn't `flags >> 4` enough) ?

Another consideration here is, will we need to change this code when we add new flags (say when we add new boolean flag `code->noinfer << 5` ? (that's why I suggested https://github.com/JuliaLang/julia/pull/42125#discussion_r702781475)

Reply:
This grabs both bits, without grabbing anything higher. So position 6 and higher are available and won't "contaminate" `noinfer` if they are nonzero.

Reply:
One stylistic side note, instead of using the integer values directly, creating an enum and using named variables makes refactoring and understanding a tad easier.

Reply:
Or maybe writing it `0b11` instead of `3`, just so it's obviously "two bits"? If you created enums, I think you'd need them for both the shifts and the masks, and that would be almost a half-dozen enums for this one `flag`. Do you think that really would be worth it?

Reply:
Ah, I get it. Thanks for your explanation.

I think I will like the idea of enum (we can do that in another PR).

Reply:
Why not create a `jl_code_info_flag_t` or something with [bit fields](https://en.cppreference.com/w/c/language/bit_field)? We use it many other places like

https://github.com/JuliaLang/julia/blob/2f00c5f6e2211ed1588976dbbe7b022148716d95/src/julia.h#L168-L175

Do we need to keep the ABI compat for Serialization.jl?

Reply:
I like that idea but [this](https://publications.gbdirect.co.uk/c_book/chapter6/bitfields.html) worries me.

Reply:
yay, it looks much cleaner

Reply:
Hmm... doesn't our GC already rely on the first field of the bitfield to be the lowerest bits?

https://github.com/JuliaLang/julia/blob/50ab3a9a422044cf3295f6f8fa24bee45a2eea1c/src/julia.h#L97-L109

i.e.., `tag->bits.gc` is used for extracting the lowest two zeros (in binary) of `tag->type`. (But yes, it'd be nice to be standard-compliant.)